pkg/kubenest/controller/virtualcluster_init_controller.go

```diff
@@ -796,28 +797,152 @@ func (c *VirtualClusterInitController) findHostAddresses() ([]string, error) {
 	if err != nil {
 		klog.Errorf("get host port pool error: %v", err)
-		return 0, err
+		return nil, err
+	}
+	next = func() nextfunc {
+		i := 0
+		return func() (int32, error) {
+			if i >= len(hostPool.PortsPool) {
+				return 0, fmt.Errorf("no available ports")
+			}
+			port := hostPool.PortsPool[i]
+			i++
+			return port, nil
+		}
+	}()
+	// }
+	return next, nil
+}
+
+// TODO: collect svc name
+func (c *VirtualClusterInitController) GetNodePorts(client kubernetes.Interface, virtualCluster *v1alpha1.VirtualCluster) ([]int32, error) {
+	ports := make([]int32, 5)
+	ipFamilies := utils.IPFamilyGenerator(constants.ApiServerServiceSubnet)
+	name := virtualCluster.GetName()
+	namespace := virtualCluster.GetNamespace()
+	apiSvc := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-%s", name, "apiserver"),
+			Namespace: namespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Type: corev1.ServiceTypeNodePort,
+			Ports: []corev1.ServicePort{
+				{
+					Port:     30007, // just for get node port
+					Protocol: corev1.ProtocolTCP,
+					TargetPort: intstr.IntOrString{
+						IntVal: 8080, // just for get node port
+					},
+					Name: "client",
+				},
+			},
+			IPFamilies: ipFamilies,
+		},
+	}
+	err := util.CreateOrUpdateService(client, apiSvc)
+	if err != nil {
+		return nil, fmt.Errorf("can not create api svc for allocate port, error: %s", err)
+	}
+
+	createdApiSvc, err := client.CoreV1().Services(namespace).Get(context.TODO(), apiSvc.GetName(), metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("can not get api svc for allocate port, error: %s", err)
+	}
+	nodePort := createdApiSvc.Spec.Ports[0].NodePort
+	ports[0] = nodePort
+
+	apiAnpAgentSvc := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-%s-konnectivity", name, "apiserver"),
+			Namespace: namespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Type: corev1.ServiceTypeNodePort,
+			Ports: []corev1.ServicePort{
+				{
+					Port:     8080, // just for get node port
+					Protocol: corev1.ProtocolTCP,
+					TargetPort: intstr.IntOrString{
+						IntVal: 8080, // just for get node port
+					},
+					Name: "serverport",
+				},
+				{
+					Port:     8081, // just for get node port
+					Protocol: corev1.ProtocolTCP,
+					TargetPort: intstr.IntOrString{
+						IntVal: 8081, // just for get node port
+					},
+					Name: "agentport",
+				},
+				{
+					Port:     8082, // just for get node port
+					Protocol: corev1.ProtocolTCP,
+					TargetPort: intstr.IntOrString{
+						IntVal: 8082, // just for get node port
+					},
+					Name: "healthport",
+				},
+				{
+					Port:     8083, // just for get node port
+					Protocol: corev1.ProtocolTCP,
+					TargetPort: intstr.IntOrString{
+						IntVal: 8083, // just for get node port
+					},
+					Name: "adminport",
+				},
+			},
+		},
+	}
```

Comment:
When rerunning, is there an issue of service update failures?

Reply:
I tried it and it won't fail